File: treessence/src/main/java/fr/bipi/tressence/ui/TextViewTree.java
package fr.bipi.tressence.ui;

import android.os.Handler;
import android.os.Looper;
import android.widget.TextView;

import org.jetbrains.annotations.NotNull;

import java.lang.ref.WeakReference;

import fr.bipi.tressence.base.FormatterPriorityTree;
import fr.bipi.tressence.formatter.Formatter;
import fr.bipi.tressence.formatter.NoTagFormatter;

/**
 * An implementation of `Timber.Tree` which sends log into a text view
 */
public class TextViewTree extends FormatterPriorityTree {

    private final UiHandler handler = new UiHandler();
    private WeakReference<TextView> textViewWeakReference = new WeakReference<>(null);

    public TextViewTree(int priority) {
        super(priority);
    }

    public void setTextView(TextView tv) {
        textViewWeakReference = new WeakReference<>(tv);
    }

    @Override
    protected Formatter getDefaultFormatter() {
        return NoTagFormatter.INSTANCE;
    }

    @Override
    protected void log(final int priority, final String tag, @NotNull final String message, Throwable t) {
        if (skipLog(priority, tag, message, t)) {
            return;
        }

        handler.post(new Runnable() {
            @Override
            public void run() {
                TextView tv = textViewWeakReference.get();
                if (tv != null) {
                    tv.setText(format(priority, tag, message));
                }
            }
        });
    }

    private static class UiHandler extends Handler {
        UiHandler() {
            super(Looper.getMainLooper());
        }
    }
}

File: treessence/src/main/java/fr/bipi/tressence/ui/TextViewTree.kt
package fr.bipi.tressence.ui

import android.os.Handler
import android.os.Looper
import android.widget.TextView
import fr.bipi.tressence.base.FormatterPriorityTree
import fr.bipi.tressence.common.filters.Filter
import fr.bipi.tressence.common.filters.NoFilter
import fr.bipi.tressence.common.formatter.Formatter
import fr.bipi.tressence.common.formatter.NoTagFormatter
import java.lang.ref.WeakReference


/**
 * An implementation of `Timber.Tree` which sends log into a text view
 */
class TextViewTree @JvmOverloads constructor(
    priority: Int,
    filter: Filter = NoFilter.INSTANCE,
    formatter: Formatter = NoTagFormatter.INSTANCE,
    private val append: Boolean = true
) : FormatterPriorityTree(
    priority, filter,
    formatter
) {
    private val handler = UiHandler()
    private var textViewWeakReference = WeakReference<TextView?>(null)

    fun setTextView(tv: TextView?) {
        textViewWeakReference = WeakReference(tv)
    }

    override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
        if (skipLog(priority, tag, message, t)) {
            return
        }
        handler.post {
            if (append) {
                textViewWeakReference.get()?.text = textViewWeakReference.get()?.text?.toString() ?: "" + format(priority, tag, message)
            } else {
                textViewWeakReference.get()?.text = format(priority, tag, message)
            }
        }
    }

    private class UiHandler : Handler(Looper.getMainLooper())
}
